python/src/iceberg/io/pyarrow.py

```diff
@@ -0,0 +1,178 @@
+    scheme, i.e. s3, hdfs, viewfs. If an existing filesystem has not been cached, it instantiates a new
+    filesystem using `pyarrow.fs.FileSystem.from_uri(location)`, caches the returned filesystem, and
+    also returns that filesystem. If a path with no scheme is provided, it's assumed to be a path to
+    a local file.
+
+    Args:
+        location(str): The location of the file
+
+    Returns:
+        pyarrow.fs.FileSystem: An implementation of the FileSystem base class inferred from the location
+
+    Raises:
+        ArrowInvalid: A suitable FileSystem implementation cannot be found based on the location provided
+    """
+    parsed_location = urlparse(location)  # Create a ParseResult from the uri
+    if not parsed_location.scheme:  # If no scheme, assume the path is to a local file
+        if _FILESYSTEM_INSTANCES.get("local"):
+            filesystem = _FILESYSTEM_INSTANCES["local"]
+        else:
+            filesystem, _ = FileSystem.from_uri(os.path.abspath(location))
+            _FILESYSTEM_INSTANCES["local"] = filesystem
+    elif _FILESYSTEM_INSTANCES.get(parsed_location.scheme):  # Check for a cached filesystem
+        filesystem = _FILESYSTEM_INSTANCES[parsed_location.scheme]
+    else:  # Instantiate a filesystem and cache it by scheme
+        filesystem, _ = FileSystem.from_uri(location)  # Infer the proper filesystem
+        _FILESYSTEM_INSTANCES[parsed_location.scheme] = filesystem  # Cache the filesystem
+    return filesystem
+
+
+class PyArrowFile(InputFile, OutputFile):
+    """A combined InputFile and OutputFile implementation that uses a pyarrow filesystem to generate pyarrow.lib.NativeFile instances
+
+    Args:
+        location(str): A URI or a path to a local file
+
+    Attributes:
+        location(str): The URI or path to a local file for a PyArrowFile instance
+        exists(bool): Whether the file exists or not
+        filesystem(pyarrow.fs.FileSystem): An implementation of the FileSystem base class inferred from the location
+
+    Examples:
+        >>> from iceberg.io.pyarrow import PyArrowFile
+        >>> input_file = PyArrowFile("s3://foo/bar.txt")
+        >>> file_content = input_file.open().read()  # Read the contents of the PyArrowFile instance
+        >>> output_file = PyArrowFile("s3://baz/qux.txt")
+        >>> output_file.create().write(b'foobytes')  # Write bytes to the PyarrowFile instance
+    """
+
+    def __init__(self, location: str):
+        self._filesystem = get_filesystem(location)  # Checks for cached filesystem for this location's scheme
+        super().__init__(location=location)
+
+    def __len__(self) -> int:
+        """Returns the total length of the file, in bytes"""
+        file = self._filesystem.open_input_file(self.location)
+        return file.size()
+
+    @property
+    def exists(self) -> bool:
+        """Checks whether the file exists"""
+        file_info = self._filesystem.get_file_info(self.location)
```

Comment:
this pattern looks strange to me, are we sure it works for non-local file systems?

Reply:
It does work for non-local file systems (at least for S3) and looks to be part of the FileSystem interface ([FileSystem.get_file_info](https://arrow.apache.org/docs/python/generated/pyarrow.fs.FileSystem.html#pyarrow.fs.FileSystem.get_file_info)). Also `type` is an attribute of the FileInfo interface ([FileInfo.type](https://arrow.apache.org/docs/python/generated/pyarrow.fs.FileInfo.html#pyarrow.fs.FileInfo.type)) that returns an enum value.

Although I did find one issue here. The full uri doesn't work but the `path` value returned from `from_uri` does, which is just the path part of the URI without the scheme or authority. I'm updating this now to store `self._filesystem, self._path = FileSystem.from_uri(location)` and using `self._path` here instead of `self.location`.